tensorflow/core/profiler/nvtx_utils.cc

```diff
@@ -0,0 +1,281 @@
+  return colors[hash % ncolor];
+}
+
+inline nvtxRangeId_t nvtxRangeStartHelper(const char* msg, const char* type,
+                                          nvtxDomainHandle_t nvtx_domain,
+                                          bool set_category = true) {
+  unsigned h = hash_string(type);
+  uint32_t color = get_color(h);
+  uint32_t category = set_category ? h : 0;
+
+  nvtxEventAttributes_t attrs = {};
+  attrs.version = NVTX_VERSION;
+  attrs.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE;
+  attrs.colorType = NVTX_COLOR_ARGB;
+  attrs.color = color;
+  attrs.messageType = NVTX_MESSAGE_TYPE_ASCII;
+  attrs.message.ascii = msg;
+  attrs.category = category;
+
+  if (nvtx_domain != NULL) return ::nvtxDomainRangeStartEx(nvtx_domain, &attrs);
+
+  return ::nvtxRangeStartEx(&attrs);
+}
+
+}  // namespace detail
+
+const NvtxDomain& GetNvtxTensorFlowCoreDomain() {
+  // Singleton because we want the same domain for the lifetime of the process.
+  static NvtxDomain nvtx_domain("tensorflow-core");
+  return nvtx_domain;
+}
+
+// A helper function to decide whether to enable CUDA NVTX profiling ranges.
+bool NvtxRangesEnabled() {
+  static bool is_enabled = [] {
+    bool _is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_ENABLE_NVTX_RANGES",
+                                               /*default_val=*/false,
+                                               &_is_enabled));
+    return _is_enabled;
+  }();
+  return is_enabled;
+}
+
+// A helper function to decide whether to enable CUDA NVTX profiling ranges
+// with detailed node information.
+bool NvtxRangesDetailedEnabled() {
+  static bool is_enabled = [] {
+    bool _is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_ENABLE_NVTX_RANGES_DETAILED",
+                                               /*default_val=*/false,
+                                               &_is_enabled));
+    return _is_enabled;
+  }();
+  return is_enabled;
+}
+
+string DataTypeToNumpyString(DataType dtype) {
+  int dtype_i = static_cast<int>(dtype);
+  bool is_ref = false;
+  if (dtype_i > 100) {
+    is_ref = true;
+    dtype_i -= 100;
+  }
+  const char* ret = "unknown";
+  // clang-format off
+  switch(dtype) {
+    case DT_INVALID: ret = "unknown"; break;
+    case DT_FLOAT: ret = "float32"; break;
+    case DT_DOUBLE: ret = "float64"; break;
+    case DT_INT32: ret = "int32"; break;
+    case DT_UINT8: ret = "uint8"; break;
+    case DT_INT16: ret = "int16"; break;
+    case DT_INT8: ret = "int8"; break;
+    case DT_STRING: ret = "string"; break;
+    case DT_COMPLEX64: ret = "complex64"; break;
+    case DT_INT64: ret = "int64"; break;
+    case DT_BOOL: ret = "bool"; break;
+    case DT_QINT8: ret = "qint8"; break;       // Not an actual Numpy type
+    case DT_QUINT8: ret = "quint8"; break;     // Not an actual Numpy type
+    case DT_QINT32: ret = "qint32"; break;     // Not an actual Numpy type
+    case DT_BFLOAT16: ret = "bfloat32";break;  // Not an actual Numpy type
```

Comment:
Should this be "bfloat16" ?

You could make this more robust by calling DataType_Name() (auto-generated by the proto compiler), absl::ConsumePrefix and absl::AsciiStrToLower.

Reply:
Indeed good point, fixed in 9595f3d588ed0115936c8e1b4c9b0ffc4413d9cd